src/plugins/procmon/procmon.cpp

```diff
@@ -128,27 +143,94 @@ static event_response_t create_user_process_hook(drakvuf_t drakvuf, drakvuf_trap
     switch ( f->format )
     {
         case OUTPUT_CSV:
-            printf("procmon," FORMAT_TIMEVAL ",%" PRIu32 ",0x%" PRIx64 ",%s,%" PRIi64",%s,%s,%s\n",
+            printf("procmon," FORMAT_TIMEVAL ",%" PRIu32 ",0x%" PRIx64 ",\"%s\",%" PRIi64",%s,0x%" PRIx64 ",%d,%s,%s\n",
                    UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.name,
-                   info->proc_data.userid, syscall_name, cmdline, imagepath);
+                   info->proc_data.userid, info->trap->name, status, new_pid, cmdline, imagepath);
             break;
 
         default:
         case OUTPUT_DEFAULT:
-            printf("[PROCMON] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 ", EPROCESS:0x%" PRIx64 ", PID:%d, PPID:%d, %s %s:%" PRIi64 " %s:%s:%s\n",
+            printf("[PROCMON] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 ", EPROCESS:0x%" PRIx64
+                   ", PID:%d, PPID:%d, \"%s\" %s:%" PRIi64 " %s:0x%" PRIx64 ":%d:%s:%s\n",
                    UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.base_addr,
                    info->proc_data.pid, info->proc_data.ppid, info->proc_data.name,
-                   USERIDSTR(drakvuf), info->proc_data.userid, syscall_name, cmdline, imagepath);
+                   USERIDSTR(drakvuf), info->proc_data.userid, info->trap->name, status, new_pid, cmdline, imagepath);
             break;
     }
 
     if (cmdline_us) vmi_free_unicode_str(cmdline_us);
     if (imagepath_us) vmi_free_unicode_str(imagepath_us);
+}
+
+static vmi_pid_t get_pid_from_handle(procmon* f, drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t handle)
+{
+    addr_t process = drakvuf_get_current_process(drakvuf, info->vcpu);
+    if (!process) return 0;
+
+    addr_t obj = drakvuf_get_obj_by_handle(drakvuf, process, handle);
+    if (!obj) return 0;
+
+    addr_t eprocess_base = obj + f->object_header_body;
+
+    vmi_pid_t pid;
+    if (VMI_FAILURE == drakvuf_get_process_pid(drakvuf, eprocess_base, &pid))
+        return 0;
+
+    return pid;
+}
+
+static event_response_t process_creation_return_hook(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    process_creation_result_t* wrapper = (process_creation_result_t*)info->trap->data;
+    if (info->regs->cr3 != wrapper->target_cr3)
+    {
+        return 0;
+    }
+
+    addr_t thread = drakvuf_get_current_thread(drakvuf, info->vcpu);
+    if (!thread || thread != wrapper->target_thread)
+    {
+        return 0;
+    }
+
+    if (info->regs->rsp != wrapper->target_rsp)
+    {
+        return 0;
+    }
+
+    procmon* f = wrapper->plugin;
+    addr_t user_process_parameters_addr = wrapper->user_process_parameters_addr;
+    addr_t new_process_handle_addr = wrapper->new_process_handle_addr;
+    reg_t status = info->regs->rax;
+
+    f->result_traps = g_slist_remove(f->result_traps, info->trap);
+    drakvuf_remove_trap(drakvuf, info->trap, drakvuf_trap_free);
```

Comment:
I'm not a fan of having a central `drakvuf_trap_free` function. I would much rather have each plugin define that for itself as the plugins may have custom structures that require more then a simple `free`.

Reply:
OK, I would use local static destructor.